From af8acffac3cec170a764615e66ae7f67c7dae6e9 Mon Sep 17 00:00:00 2001
From: Chad Dupuis <cdupuis@redhat.com>
Date: Mon, 23 Jun 2014 20:01:08 -0400
Subject: [scsi] qla2xxx: Don't check for firmware hung during the reset context for ISP82XX

Message-id: <1403553672-3844-3-git-send-email-cdupuis@redhat.com>
Patchwork-id: 84410
O-Subject: [PATCH RHEL 7.1 e-stor 2/6] qla2xxx: Don't check for firmware hung during the reset context for ISP82XX.
Bugzilla: 1090378
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Rob Evers <revers@redhat.com>

From: Tej Prakash <tej.prakash@qlogic.com>

scsi commit id 03d32f978a758862b6e6ba228d3e28f6dfe9ec68

Signed-off-by: Tej Prakash <tej.prakash@qlogic.com>
Signed-off-by: Saurav Kashyap <saurav.kashyap@qlogic.com>
Signed-off-by: Christoph Hellwig <hch@lst.de>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/scsi/qla2xxx/qla_nx.c b/drivers/scsi/qla2xxx/qla_nx.c
index b850138..4822533 100644
--- a/drivers/scsi/qla2xxx/qla_nx.c
+++ b/drivers/scsi/qla2xxx/qla_nx.c
@@ -2826,47 +2826,30 @@ static int
 qla82xx_device_bootstrap(scsi_qla_host_t *vha)
 {
  int rval = QLA_SUCCESS;
- int i, timeout;
+ int i;
  uint32_t old_count, count;
  struct qla_hw_data *ha = vha->hw;
- int need_reset = 0, peg_stuck = 1;
+ int need_reset = 0;
 
  need_reset = qla82xx_need_reset(ha);
 
- old_count = qla82xx_rd_32(ha, QLA82XX_PEG_ALIVE_COUNTER);
-
- for (i = 0; i < 10; i++) {
-  timeout = msleep_interruptible(200);
-  if (timeout) {
-   qla82xx_wr_32(ha, QLA82XX_CRB_DEV_STATE,
-    QLA8XXX_DEV_FAILED);
-   return QLA_FUNCTION_FAILED;
-  }
-
-  count = qla82xx_rd_32(ha, QLA82XX_PEG_ALIVE_COUNTER);
-  if (count != old_count)
-   peg_stuck = 0;
- }
-
  if (need_reset) {
   /* We are trying to perform a recovery here. */
-  if (peg_stuck)
+  if (ha->flags.isp82xx_fw_hung)
    qla82xx_rom_lock_recovery(ha);
-  goto dev_initialize;
  } else  {
-  /* Start of day for this ha context. */
-  if (peg_stuck) {
-   /* Either we are the first or recovery in progress. */
-   qla82xx_rom_lock_recovery(ha);
-   goto dev_initialize;
-  } else
-   /* Firmware already running. */
-   goto dev_ready;
+  old_count = qla82xx_rd_32(ha, QLA82XX_PEG_ALIVE_COUNTER);
+  for (i = 0; i < 10; i++) {
+   msleep(200);
+   count = qla82xx_rd_32(ha, QLA82XX_PEG_ALIVE_COUNTER);
+   if (count != old_count) {
+    rval = QLA_SUCCESS;
+    goto dev_ready;
+   }
+  }
+  qla82xx_rom_lock_recovery(ha);
  }
 
- return rval;
-
-dev_initialize:
  /* set to DEV_INITIALIZING */
  ql_log(ql_log_info, vha, 0x009e,
      "HW State: INITIALIZING.\n");
-- 
1.7.1